### ISSUE TITLE
[TEST] Create integration testing for SPDK engine

### ISSUE THREAD

**Issue by innobead** [longhorn/longhorn#5834]
## What's the test to develop? Please describe

Need to have component-level integration testing for https://github.com/longhorn/longhorn-spdk-engine. We probably can leverage the existing longhorn-engine testing w/ some adjustments. 

## Describe the items of the test development (DoD, definition of done) you'd like

- [ ] Have fundamental integration testing for SPDK-engine, and run for PR or image push via tag.

## Additional context

cc @longhorn/dev-data-plane

**Comment by shuo-wu** [comment]
The integration test in longhorn/longhorn-spdk-engine is already there. To enable it, we still need to have an env set-up: https://github.com/longhorn/longhorn-spdk-engine/blob/main/pkg/spdk_test.go

As for the test cases in longhorn/longhorn-tests, we will see if we can improve the existing cases first so that they can fit both v1 and v2 volumes. Otherwise, we will create new cases exclusively for v2.

**Comment by shuo-wu** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Longhorn-Sprint-59d8105b2fbdf4155c4589ef/issues/longhorn/longhorn/5834?planning-poker) with Zenhub @DamiaSan @derekbit @innobead @chriscchien @yangchiu

**Comment by DamiaSan** [comment]
> The integration test in longhorn/longhorn-spdk-engine is already there. To enable it, we still need to have an env set-up:

Can you describe better the work to do?

**Comment by shuo-wu** [comment]
Currently, [longhorn/longhorn-spdk-engine ci](https://drone-pr.longhorn.io/longhorn/longhorn-spdk-engine) build section will run the test I mentioned but get failed. Since we haven't set up the test env for it. 
The reason why it won't block the merge is, we [skip the result validation](https://github.com/longhorn/longhorn-spdk-engine/blob/c9046e18cf8c106ce403294c6153e5057bffae1a/scripts/ci#L9-L12)

**Comment by DamiaSan** [comment]
An issue that I encounter running `longhorn-spdk-engine` test is the following:
```
FAIL: spdk_test.go:160: TestSuite.TestSPDKMultipleThread

spdk_test.go:78:
    c.Assert(err, IsNil)
... value *errors.fundamental = timeout executing: /usr/bin/nsenter [nsenter --mount=/proc/1/ns/mnt --ipc=/proc/1/ns/ipc --net=/proc/1/ns/net sh -c /usr/src/spdk/build/bin/spdk_tgt  --logflag all 2>&1 | tee /tmp/spdk_tgt.log] ("timeout executing: /usr/bin/nsenter [nsenter --mount=/proc/1/ns/mnt --ipc=/proc/1/ns/ipc --net=/proc/1/ns/net sh -c /usr/src/spdk/build/bin/spdk_tgt  --logflag all 2>&1 | tee /tmp/spdk_tgt.log]")
```
The command to start `spdk_tgt` goes into timeout, but the tgt is effectively started and so the test can continue. All the following checks are ok, but the test shows a failed state due to the timeout above.
This error does not happen if I run the tgt and execute the test manually.

**Comment by DamiaSan** [comment]
On older version of the kernel (5.15.0) the test return also the following errors:
```
spdk_test.go:249:
    c.Assert(engine.State, Equals, types.InstanceStateRunning)
... obtained string = "error"
... expected string = "running"

spdk_test.go:512:
    c.Assert(engine.State, Equals, types.InstanceStateRunning)
... obtained string = "error"
... expected string = "running"
```
Also there errors do not happen if the test and the tgt are started manually.

**Comment by DamiaSan** [comment]
Instead, the following error happen also with the test and the tgt started manually:
```
time="2024-03-21T11:31:50Z" level=error msg="spdk gRPC server: failed to verify and update replica cache, will retry later" error="error sending message, id 0, method bdev_get_bdevs, params <nil>: context done during async message send"
```
But this does not invalidate the test result.

**Comment by shuo-wu** [comment]
I don't think we config something like huge pages for the ci env. Is the spdk_tgt runnable?

**Comment by DamiaSan** [comment]
> I don't think we config something like huge pages for the ci env. Is the spdk_tgt runnable?

No, without huge pages configuration, spdk_tgt doesn't start.

However, what I wanted to say with my above comments is that, before to put again the test under CI, we should resolve the "false" errors I described.
